### ISSUE TITLE
Unable to start server: block does not exist

### ISSUE THREAD

**Issue by jrick** [decred/dcrd#2643]
This may have been due to unclean shutdown, but it doesn't appear to be database corruption (at least that leveldb can determine)

```
$ dcrd --testnet -dtrace 
2021-05-04 09:15:23.808 [INF] DCRD: Version 1.7.0-pre (Go version go1.16.3 openbsd/amd64)
2021-05-04 09:15:23.808 [INF] DCRD: Home dir: /home/jrick/.dcrd
2021-05-04 09:15:23.808 [INF] DCRD: Loading block database from '/home/jrick/.dcrd/data/testnet3/blocks_ffldb'
2021-05-04 09:15:23.943 [TRC] BCDB: Scan found latest block file #12 with length 94031447
2021-05-04 09:15:23.945 [INF] DCRD: Block database loaded
2021-05-04 09:15:23.945 [INF] CHAN: Loading UTXO database from '/home/jrick/.dcrd/data/testnet3/utxodb'
2021-05-04 09:15:24.002 [TRC] BCDB: Scan found latest block file #-1 with length 0
2021-05-04 09:15:24.003 [INF] CHAN: UTXO database loaded
2021-05-04 09:15:24.029 [INF] INDX: Exists address index is enabled
2021-05-04 09:15:24.063 [DBG] FEES: Loaded fee estimator database
2021-05-04 09:15:24.093 [TRC] CHAN: Serialized chain state: 1770cb789db90fd83e6d9b808f7be7f6194905f7e316dd6a6132ae985d000000494b0a00107290000000000069bab18715bd010008000000e4195c08a9a4b022
2021-05-04 09:15:24.093 [INF] CHAN: Loading block index...
2021-05-04 09:15:30.569 [DBG] CHAN: Block index loaded in 6.475967693s
2021-05-04 09:15:30.906 [INF] STKE: Stake database version 1 loaded
2021-05-04 09:15:30.906 [DBG] CHAN: Most recent checkpoint is 00000010ecddf8da5d91f7020f69130db8a163906d460cbbed2a91568701f0ac (height 515730)
2021-05-04 09:15:30.909 [INF] CHAN: UTXO cache initializing (max size: 150 MiB)...
2021-05-04 09:15:30.910 [ERR] DCRD: Unable to start server: block 0000004ecbbd968c04abd0ec8f71c0733a5214fe39b1ea097a9ed3459fe61ae7 does not exist
2021-05-04 09:15:30.910 [INF] DCRD: Gracefully shutting down the UTXO database...
2021-05-04 09:15:30.910 [INF] DCRD: Gracefully shutting down the block database...
2021-05-04 09:15:30.914 [INF] DCRD: Shutdown complete
$ echo $?
1
```

Perhaps of note:

2021-05-04 09:15:24.002 [TRC] BCDB: Scan found latest block file #-1 with length 0

seems to indicate that there is no latest block file at all, but this seems to be an expected case as the caller checks for the file -1.

**Comment by jrick** [comment]
The log ends abruptly, so this is definitely a case of unclean shutdown.

```
...
2021-05-01 13:18:02.784 [INF] SYNC: New block 00000035c253939b3d678bc3c44aeebda0bbabd8f0f35b7384fdcfe719c5b586 (1 transaction, 9 tickets, 5 votes, 0 revocations, height 674629, interval 2m0s)
2021-05-01 13:18:51.207 [INF] SYNC: New block 00000055c752c566ef1135f9f467a36abf2fe6bc34f3610b1b07b6a438736b1a (9 transactions, 0 tickets, 5 votes, 0 revocations, height 674630, interval 51s)
2021-05-01 13:19:39.562 [INF] SYNC: New block 0000001784564b56916f7d457f0f4cc3839b9455397c46ecdbe4206550b119d9 (4 transactions, 13 tickets, 5 votes, 0 revocations, height 674631, interval 54s)
2021-05-01 13:21:04.896 [INF] SYNC: New block 00000011d97a88a866aaaa1ab15991c298eafe50b77727374d8c590859aa5216 (3 transactions, 1 ticket, 5 votes, 0 revocations, height 674632, interval 1m15s)
2021-05-01 13:22:50.964 [INF] SYNC: New valid peer [2001:41d0:801:2000::8b2]:60834 (inbound) (/dcrwire:0.4.0/decred-mapper:0.0.1/)
2021-05-01 13:23:08.139 [INF] SYNC: New block 0000005d98ae32616add16e3f7054919f6e77b8f809b6d3ed80fb99d78cb7017 (2 transactions, 1 ticket, 5 votes, 0 revocations, height 674633, interval 2m0s)
2021-05-01 13:24:46.099 [INF] SYNC: New block 0000004ecbbd968c04abd0ec8f71c0733a5214fe39b1ea097a9ed3459fe61ae7 (10 transactions, 2 tickets, 5 votes, 0 revocations, height 674634, interval 1m45s)
2021-05-01 13:25:46.690 [INF] SYNC: New block 00000039f2517498c2fe1907e9be2f3ae36238c7fe4f97d5c0072115f70dc6c2 (2 transactions, 11 tickets, 5 votes, 0 revocations, height 674635, interval 1m0s)
2021-05-01 13:26:06.536 [INF] SYNC: New valid peer [2001:41d0:801:2000::8b2]:35852 (inbound) (/dcrwire:0.4.0/dcrseeder:0.0.1/)
2021-05-01 13:26:30.902 [INF] SYNC: New block 000000526d2d27ab78e2eaab03038321c41811ebba1cbaf00611a268b27e2318 (2 transactions, 0 tickets, 5 votes, 0 revocations, height 674636, interval 45s)
```

**Comment by rstaudt2** [comment]
I was able to reproduce this, and it was introduced by moving the UTXO set into its own database.  The issue is that the `ffldb` database driver has its own cache that only flushes to disk every 5 minutes if the cache is not full.  Now that the block and UTXO databases are separate, the databases will end up flushing to disk at different times.

In this case, it looks like the utxo database flushed to disk at some point after block 674634, but the block database had not.  Thus the error of the UTXO cache unexpectedly not being able to find that block on startup. 

I was able to recreate this more easily by adding a bit of code to force the UTXO database to flush immediately but kept the block database at 5 minutes so that there is a large window of time to kill `dcrd` and force an unclean shutdown.  I'll look into a fix to handle this case properly.
